pkg/maps/ctmap/ctmap.go

```diff
@@ -513,95 +412,84 @@ func purgeCtEntry4(m *Map, key CtKey, entry *CtEntry, natMap *nat.Map, next func
 	return nil
 }
 
-// doGC4 iterates through a CTv4 map and drops entries based on the given
-// filter.
-func doGC4(m *Map, filter GCFilter, next func(GCEvent)) gcStats {
-	var natMap *nat.Map
+var (
+	batchAPISupported     bool
+	batchAPISupportedOnce sync.Once
+)
 
-	if m.clusterID == 0 {
-		// global map handling
-		ctMap := mapInfo[m.mapType]
-		if ctMap.natMapLock != nil {
-			ctMap.natMapLock.Lock()
-			defer ctMap.natMapLock.Unlock()
-		}
-		natMap = ctMap.natMap
-	} else {
-		// per-cluster map handling
-		natm, err := nat.GetClusterNATMap(m.clusterID, nat.IPv4)
-		if err != nil {
-			log.WithError(err).Error("Unable to get per-cluster NAT map")
-		} else {
-			natMap = natm
+func isBatchAPISupported() bool {
+	batchAPISupportedOnce.Do(func() {
+		err := probes.HaveBatchAPI()
+		batchAPISupported = !errors.Is(err, probes.ErrNotSupported)
+		if !batchAPISupported {
+			log.WithError(err).Info("kernel does not support batch iteration for ctmap gc")
 		}
+	})
+	return batchAPISupported
+}
+
+func iterate[KT any, VT any, KP bpf.KeyPointer[KT], VP bpf.ValuePointer[VT]](m *Map, stats *gcStats, filterCallback func(key bpf.MapKey, value bpf.MapValue)) error {
+	// Note: We can drop this once the minimum supported kernel version has batch iteration (i.e. >=5.6).
+	if !isBatchAPISupported() {
+		return m.DumpReliablyWithCallback(filterCallback, stats.DumpStats)
 	}
 
-	stats := statStartGc(m)
-	defer stats.finish()
+	ctx := context.Background()
+	iter := bpf.NewBatchIterator[KT, VT, KP, VP](&m.Map)
+	for k, v := range iter.IterateAll(ctx) {
+		filterCallback(k, v)
+	}
+	stats.Completed = true
+	return iter.Err()
+}
 
-	if natMap != nil {
-		if err := natMap.Open(); err == nil {
-			defer natMap.Close()
-		} else {
-			natMap = nil
+var _ tupleKeyAccessor = &tuple.TupleKey4{}
+var _ tupleKeyAccessor = &tuple.TupleKey6{}
+
+type tupleKeyAccessor interface {
+	GetDestAddr() netip.Addr
+	GetDestPort() uint16
+	GetSourceAddr() netip.Addr
+	GetSourcePort() uint16
+	GetNextHeader() u8proto.U8proto
+	GetFlags() uint8
+}
+
+func cleanup[KT any, VT any, KP bpf.KeyPointer[KT], VP bpf.ValuePointer[VT]](m *Map, filter GCFilter, natMap *nat.Map, stats *gcStats, next func(GCEvent), ipv6 bool) func(key bpf.MapKey, value bpf.MapValue) {
```

Comment:
It seems that the type parameter of the cleanup function is not being referenced.